internal/runners/builds/download.go

```diff
@@ -0,0 +1,176 @@
+	}
+
+	_, err := runtime.NewFromProject(d.project, target.TriggerBuilds, d.analytics, d.svcModel, d.out, d.auth, d.config)
+	if err != nil {
+		return locale.WrapInputError(err, "err_refresh_runtime_new", "Could not update runtime for this project.")
+	}
+
+	runtimeStore := store.New(target.NewProjectTarget(d.project, nil, target.TriggerBuilds).Dir())
+	bp, err := runtimeStore.BuildPlan()
+	if err != nil {
+		return errs.Wrap(err, "Could not get build plan")
+	}
+
+	terminalArtfMap, err := buildplan.NewMapFromBuildPlan(bp, false, false, nil)
+	if err != nil {
+		return errs.Wrap(err, "Could not get build plan map")
+	}
+
+	var artifact *artifact.Artifact
+	for _, artifacts := range terminalArtfMap {
+		for _, a := range artifacts {
+			if strings.HasPrefix(strings.ToLower(string(a.ArtifactID)), strings.ToLower(params.BuildID)) {
+				artifact = &a
+				break
+			}
+		}
+	}
+
+	if artifact == nil {
+		return locale.NewInputError("err_build_id_not_found", "Could not find artifact with ID {{.V0}}", params.BuildID)
+	}
+
+	targetDir := params.OutputDir
+	if targetDir == "" {
+		targetDir, err = os.Getwd()
+		if err != nil {
+			return errs.Wrap(err, "Could not get current working directory")
+		}
+	}
+
+	if err := d.downloadArtifact(artifact, targetDir); err != nil {
+		return errs.Wrap(err, "Could not download artifact %s", artifact.ArtifactID.String())
+	}
+
+	return nil
+}
+
+func (d *Download) downloadArtifact(artifact *artifact.Artifact, targetDir string) (rerr error) {
+	ctx, cancel := context.WithCancel(context.Background())
+	var w io.Writer = os.Stdout
+	if d.out.Type() != output.PlainFormatName {
+		w = nil
+	}
+
+	pg := mpb.NewWithContext(
+		ctx,
+		mpb.WithOutput(w),
+		mpb.WithWidth(40),
+		mpb.WithRefreshRate(constants.TerminalAnimationInterval),
+	)
+	defer cancel()
+
+	name := artifact.Name
+	if len(name) > progress.MaxNameWidth() {
+		name = name[0:progress.MaxNameWidth()]
+	}
+
+	prependDecorators := []decor.Decorator{
+		decor.Name(name, decor.WC{W: progress.MaxNameWidth(), C: decor.DidentRight}),
+		decor.OnComplete(
+			decor.Spinner(progress.SpinnerFrames, decor.WCSyncSpace), "",
+		),
+		decor.CountersKiloByte("%.1f/%.1f", decor.WC{W: 17}),
+	}
+
+	options := []mpb.BarOption{
+		mpb.BarFillerClearOnComplete(),
+		mpb.PrependDecorators(prependDecorators...),
+		mpb.AppendDecorators(
+			decor.OnComplete(decor.Percentage(decor.WC{W: 5}), ""),
+		),
+	}
+
+	artifactURL, err := url.Parse(artifact.URL)
+	if err != nil {
+		return errs.Wrap(err, "Could not parse artifact URL %s.", artifact.URL)
+	}
+
+	var downloadBar *mpb.Bar
+	b, err := httputil.GetWithProgress(artifactURL.String(), &rtProgress.Report{
+		ReportSizeCb: func(size int) error {
+			downloadBar = pg.AddBar(int64(size), options...)
+			return nil
+		},
+		ReportIncrementCb: func(inc int) error {
+			downloadBar.IncrBy(inc)
+			return nil
+		},
+	})
+	if err != nil {
+		return errs.Wrap(err, "Download %s failed", artifactURL.String())
+	}
+
```

Comment:
You'll need to wait for the progressbar to be done, the reporting is done in a subroutine, and right now your function ends before it has completed:

```
state builds dl AA0F839B
█ Download build artifacts (Beta)

Beta Feature: This feature is still in beta and may be unstable.

bzip2                          |    8.4KB/117.6KB [==>-----------------------------------]   7 %
Downloaded bzip2 to /Users/nathanrijksen/Projects/cli/artifact.tar.gz
```

Reply:
Added it and it's gross